python/cudf/cudf/core/join/join.py

```diff
@@ -0,0 +1,454 @@
+        same_named_columns = set(lhs._data.keys()) & set(rhs._data.keys())
+        if (
+            not (left_index or right_index)
+            and not (left_on or right_on)
+            and len(same_named_columns) == 0
+        ):
+            raise ValueError("No common columns to perform merge on")
+
+        if suffixes:
+            lsuffix, rsuffix = suffixes
+        for name in same_named_columns:
+            if name == left_on == right_on:
+                continue
+            elif left_on and right_on:
+                if (name in left_on and name in right_on) and (
+                    left_on.index(name) == right_on.index(name)
+                ):
+                    continue
+            else:
+                if not (lsuffix or rsuffix):
+                    raise ValueError(
+                        "there are overlapping columns but "
+                        "lsuffix and rsuffix are not defined"
+                    )
+
+    def typecast_input_to_libcudf(self):
+
+        lhs_keys, rhs_keys, lhs_cols, rhs_cols = [], [], [], []
+        if self.left_index:
+            lhs_keys.append(self.lhs.index._data.keys())
+            lhs_cols.append(self.lhs.index)
+        if self.right_index:
+            rhs_keys.append(self.rhs.index._data.keys())
+            rhs_cols.append(self.rhs.index)
+        if self.left_on:
+            lhs_keys.append(self.left_on)
+            lhs_cols.append(self.lhs)
+        if self.right_on:
+            rhs_keys.append(self.right_on)
+            rhs_cols.append(self.rhs)
+
+        for l_key_grp, r_key_grp, l_col_grp, r_col_grp in zip(
+            lhs_keys, rhs_keys, lhs_cols, rhs_cols
+        ):
+            for l_key, r_key in zip(l_key_grp, r_key_grp):
+                to_dtype = self.input_to_libcudf_casting_rules(
+                    l_col_grp._data[l_key], r_col_grp._data[r_key], self.how
+                )
+                l_col_grp._data[l_key] = l_col_grp._data[l_key].astype(
+                    to_dtype
+                )
+                r_col_grp._data[r_key] = r_col_grp._data[r_key].astype(
+                    to_dtype
+                )
+
+    def input_to_libcudf_casting_rules(self, lcol, rcol, how):
+        cast_warn = "can't safely cast column {} from {} with type \
+                        {} to {}, upcasting to {}"
+        ctgry_err = "can't implicitly cast column {0} to categories \
+                        from {1} during {1} join"
+
+        dtype_l = lcol.dtype
+        dtype_r = rcol.dtype
+        libcudf_join_type = None
+        if pd.api.types.is_dtype_equal(dtype_l, dtype_r):
+            # if categorical and equal, children passed to libcudf
+            libcudf_join_type = dtype_l
+        elif is_categorical_dtype(dtype_l) and is_categorical_dtype(dtype_r):
+            # categories are not equal
+            libcudf_join_type = np.dtype("O")
+        elif how == "left":
+            check_col = rcol.fillna(0)
```

Comment:
I believe this is because `NumericalColumn.overflow_safe_to` might use a cast and a binop to check for equality with the casted column, followed by an `all()`, and we don't want NaNs that don't compare as each other as equal to mess up the condition if all the other values do compare as equal. However I think that only happens in a subset of cases so it probably makes sense to migrate the fillna to the overflow checking function.

Reply:
Moved